Skip setting type expansion context for captures of skipped functions

jkshtj opened the pull request:
Type-checking and SILGen for non-inlinable functions is skipped in the presence of `-experimental-skip-non-inlinable-function-bodies` and `-experimental-skip-non-inlinable-function-bodies-without-types` flags.

Such functions may be top-level and may contain captures (if they appear after a `guard` statement), for which we were setting the type expansion context during SILGen. Setting type expansion context however, relied on the capture info being computed -- which was never happening because of the abovementioned flags.

Changes in this commit make setting type expansion context, for captures, conditional on a flag that ensures that the function has already been typechecked.

Fixes #57646

jkshtj commented:
@BradLarson could you please kick off swift-ci?

slavapestov commented:
> This means that such functions should be typechecked and SILGen-ed in the presence of -experimental-skip-non-inlinable-function-bodies and -experimental-skip-non-inlinable-function-bodies-without-types flags.

I’m not sure I understand why. These functions cannot be referenced so why do we need their captures when skipping bodies?

jkshtj commented:
> > This means that such functions should be typechecked and SILGen-ed in the presence of -experimental-skip-non-inlinable-function-bodies and -experimental-skip-non-inlinable-function-bodies-without-types flags.
> 
> I’m not sure I understand why. These functions cannot be referenced so why do we need their captures when skipping bodies?

You're right. I was initially going off of the function SILGen invariants you established in [this](https://github.com/apple/swift/pull/28596/commits/8214e18850e80847eb28e04880ea689f8e076e99) commit. But if a function's body is never going to be emitted then I don't see any reason why would need their captures. 

One incorrect assumption that I made initially was that this capture info is represented in the "function signatures" on top of a SIL file. But, those are swift signatures.

I think the fix should be to simply not set the type expansion context for captures [here](https://github.com/apple/swift/blob/main/lib/SILGen/SILGen.cpp#L1440) based on whether or not the function body is going to be emitted?

jkshtj committed:
Skip setting type expansion context for captures of skipped functions

Type-checking and SILGen for non-inlinable functions is skipped in the
presence of `-experimental-skip-non-inlinable-function-bodies` and
`-experimental-skip-non-inlinable-function-bodies-without-types` flags.

Such functions may be top-level and may contain captures (if they appear
after a `guard` statement), for which we were setting the type expansion
context during SILGen. Setting type expansion context however, relied on
the capture info being computed -- which was never happening because of
the abovementioned flags.

Changes in this commit make setting type expansion context, for
captures, conditional on a flag that ensures that the function has already
been typechecked.

Fixes #57646

jkshtj commented:
@slavapestov could you please take a look at this?

BradLarson commented:
@swift-ci Please test

slavapestov reviewed (approved):
Nice!

jkshtj commented:
@slavapestov Is there anything else needed to merge this? Or are we waiting on more approvals?
